cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/TransformExecutorFactory.java

```diff
@@ -102,69 +101,68 @@ protected abstract TrackedTransform getTransformation(String pluginType, String
+    return new PipeTransformExecutor<>(transformations, startingPoints);
   }
 
-  private <KEY_OUT, VAL_OUT> void setBatchTransformDetail(PipelinePhase pipeline, String stageName,
-                                                          Map<String, BatchTransformDetail> transformations,
-                                                          Map<String, ErrorOutputWriter<Object, Object>>
-                                                            transformErrorSinkMap,
-                                                          OutputWriter<KEY_OUT, VAL_OUT> outputWriter)
+  private <KEY_OUT, VAL_OUT> void setPipeTransformDetail(PipelinePhase pipeline, String stageName,
+                                                         Map<String, PipeTransformDetail> transformations,
+                                                         Map<String, ErrorOutputWriter<Object, Object>>
+                                                           transformErrorSinkMap,
+                                                         OutputWriter<KEY_OUT, VAL_OUT> outputWriter)
     throws Exception {
     if (pipeline.getSinks().contains(stageName)) {
-      String pluginType = pipeline.getStage(stageName).getPluginType();
       // If there is a connector sink/ joiner at the end of pipeline, do not remove stage name. This is needed to save
       // stageName along with the record in connector sink and joiner takes input along with stageName
-      boolean removeStageName = !(pluginType.equalsIgnoreCase(Constants.CONNECTOR_TYPE) ||
-        pluginType.equalsIgnoreCase(BatchJoiner.PLUGIN_TYPE));
-      transformations.put(stageName, new BatchTransformDetail(stageName, removeStageName,
-                                                              getTransformation(pluginType, stageName),
-                                                              new SinkEmitter<>(stageName, outputWriter)));
+      String pluginType = pipeline.getStage(stageName).getPluginType();
+      boolean removeStageName = !(pluginType.equals(Constants.CONNECTOR_TYPE) ||
+        pluginType.equals(BatchJoiner.PLUGIN_TYPE));
+      transformations.put(stageName, new PipeTransformDetail(stageName, removeStageName,
+                                                             getTransformation(pluginType, stageName),
+                                                             new SinkEmitter<>(stageName, outputWriter)));
       return;
     }
 
     for (String output : pipeline.getDag().getNodeOutputs(stageName)) {
-      setBatchTransformDetail(pipeline, output, transformations, transformErrorSinkMap, outputWriter);
+      setPipeTransformDetail(pipeline, output, transformations, transformErrorSinkMap, outputWriter);
 
-      // Add new transformation for output stage if the transfomation for stageName already exists
+      // Add new transformation for output stage if the transformation for stageName already exists
       if (transformations.containsKey(stageName)) {
```

Comment:
Ok I see it's because the emitter constructors take a map of outputs. It would be simpler to understand if you remove the map from the constructors and move the logic to put the PipelineTransformDetail for this stageName before this for loop. The loop then just performs a recursive call and adds the output for this stageName.